src/context.rs

```diff
@@ -312,6 +420,374 @@ const REF_TYPES: usize = 2;
+const SIG_COEF_CONTEXTS_1D: usize =  16;
+const SIG_COEF_CONTEXTS_EOB: usize =  4;
+const SIG_COEF_CONTEXTS: usize = SIG_COEF_CONTEXTS_2D + SIG_COEF_CONTEXTS_1D;
+
+const COEFF_BASE_CONTEXTS: usize = SIG_COEF_CONTEXTS;
+const DC_SIGN_CONTEXTS: usize =  3;
+
+const BR_TMP_OFFSET: usize =  12;
+const BR_REF_CAT: usize =  4;
+const LEVEL_CONTEXTS: usize =  21;
+
+const NUM_BASE_LEVELS: usize =  2;
+
+const BR_CDF_SIZE: usize = 4;
+const COEFF_BASE_RANGE: usize = 4 * (BR_CDF_SIZE - 1);
+
+const COEFF_CONTEXT_BITS: usize = 6;
+const COEFF_CONTEXT_MASK: usize = (1 << COEFF_CONTEXT_BITS) - 1;
+const MAX_BASE_BR_RANGE: usize = COEFF_BASE_RANGE + NUM_BASE_LEVELS + 1;
+
+const BASE_CONTEXT_POSITION_NUM: usize = 12;
+
+// Pad 4 extra columns to remove horizontal availability check.
+const TX_PAD_HOR_LOG2: usize = 2;
+const TX_PAD_HOR: usize = 4;
+// Pad 6 extra rows (2 on top and 4 on bottom) to remove vertical availability
+// check.
+const TX_PAD_TOP: usize = 2;
+const TX_PAD_BOTTOM: usize = 4;
+const TX_PAD_VER: usize = (TX_PAD_TOP + TX_PAD_BOTTOM);
+// Pad 16 extra bytes to avoid reading overflow in SIMD optimization.
+const TX_PAD_END: usize = 16;
+const TX_PAD_2D: usize = ((MAX_TX_SIZE + TX_PAD_HOR) * (MAX_TX_SIZE + TX_PAD_VER) + TX_PAD_END);
+
+const TX_CLASSES: usize = 3;
+
+#[derive(Copy,Clone,PartialEq)]
+pub enum TxClass {
+  TX_CLASS_2D = 0,
+  TX_CLASS_HORIZ = 1,
+  TX_CLASS_VERT = 2,
+}
+
+use context::TxClass::*;
+
+static tx_type_to_class: [TxClass; TX_TYPES] = [
+    TX_CLASS_2D,     // DCT_DCT
+    TX_CLASS_2D,     // ADST_DCT
+    TX_CLASS_2D,     // DCT_ADST
+    TX_CLASS_2D,     // ADST_ADST
+    TX_CLASS_2D,     // FLIPADST_DCT
+    TX_CLASS_2D,     // DCT_FLIPADST
+    TX_CLASS_2D,     // FLIPADST_FLIPADST
+    TX_CLASS_2D,     // ADST_FLIPADST
+    TX_CLASS_2D,     // FLIPADST_ADST
+    TX_CLASS_2D,     // IDTX
+    TX_CLASS_VERT,   // V_DCT
+    TX_CLASS_HORIZ,  // H_DCT
+    TX_CLASS_VERT,   // V_ADST
+    TX_CLASS_HORIZ,  // H_ADST
+    TX_CLASS_VERT,   // V_FLIPADST
+    TX_CLASS_HORIZ,  // H_FLIPADST
+];
+
+static eob_to_pos_small: [u8; 33] = [
+    0, 1, 2,                                        // 0-2
+    3, 3,                                           // 3-4
+    4, 4, 4, 4,                                     // 5-8
+    5, 5, 5, 5, 5, 5, 5, 5,                         // 9-16
+    6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6  // 17-32
+];
+
+static eob_to_pos_large: [u8; 17] = [
+    6,                               // place holder
+    7,                               // 33-64
+    8,  8,                           // 65-128
+    9,  9,  9,  9,                   // 129-256
+    10, 10, 10, 10, 10, 10, 10, 10,  // 257-512
+    11                               // 513-
+];
+
+
+static k_eob_group_start: [u16; 12] = [ 0, 1, 2, 3, 5, 9,
+                                        17, 33, 65, 129, 257, 513 ];
+static k_eob_offset_bits: [u16; 12] = [ 0, 0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 ];
+
+static clip_max3: [u8; 256] = [
```

Comment:
I know this is probably how libaom does this clip... but I really doubt using a lookup table is faster than a normal clip here.